lib/storage/clusterconfig/kubernetes_test.go

```diff
@@ -0,0 +1,328 @@
+					"cloud.google.com/load-balancer-type": "Internal",
+				},
+				Spec: ControllerServiceSpec{
+					Type: NodePort,
+				},
+			},
+			expectedServiceConfig: &GravityControllerService{
+				Labels: map[string]string{
+					"app-existing": constants.GravityServiceName,
+				},
+				Annotations: map[string]string{
+					"cloud.google.com/load-balancer-type": "Internal",
+				},
+				Spec: ControllerServiceSpec{
+					Type: NodePort,
+					Ports: []Port{
+						{
+							Name:     "port-existing",
+							Port:     3001,
+							NodePort: 32001,
+						},
+					},
+				},
+			},
+			comment: "maintain existing spec if undefined in updated service config",
+		},
+	}
+	for _, tc := range testCases {
+		comment := Commentf(tc.comment)
+
+		// clusterControl provides control over updated service config.
+		clusterControl := r.NewClusterConfigControl(tc.updatedServiceConfig)
+
+		// serviceControl provides control over existing service.
+		serviceControl := r.NewServiceControl(tc.existingServiceConfig)
+
+		c.Assert(Reconcile(clusterControl, serviceControl), IsNil, comment)
+
+		serviceConfig, err := serviceControl.Get()
+		c.Assert(err, IsNil, comment)
+		c.Assert(serviceConfig, compare.DeepEquals, tc.expectedServiceConfig, comment)
+	}
+}
+
+// mockClusterConfigControl provides mock implementation of ClusterConfigControl.
+type mockClusterConfigControl struct {
+	resource Resource
+}
+
+// NewClusterConfigControl returns a new cluster config control for the provided
+// config.
+func (r *KubernetesSuite) NewClusterConfigControl(config *GravityControllerService) *mockClusterConfigControl {
+	return &mockClusterConfigControl{
+		resource: Resource{
+			Spec: Spec{
+				ComponentConfigs: ComponentConfigs{
+					GravityControllerService: config,
+				},
+			},
+		},
+	}
+}
+
+// Get returns the cluster's ClusterConfiguration resource.
+func (r *mockClusterConfigControl) Get() (*Resource, error) {
+	return &r.resource, nil
+}
+
+// Update updates the cluster's ClusterConfiguration resource.
+func (r *mockClusterConfigControl) Update(_ *Resource) error {
+	return trace.NotImplemented("not implemented for mockClusterConfigControl")
+}
+
+// mockServiceControl provides mock implementation of ServiceControl.
+type mockServiceControl struct {
+	svc *v1.Service
+}
+
+// NewServiceControl returns a new service control constructed from the provided
+// config.
+func (r *KubernetesSuite) NewServiceControl(config *GravityControllerService) *mockServiceControl {
+	return &mockServiceControl{
+		svc: newService(config),
+	}
+}
+
+// Get returns the controller service configuration.
+func (r *mockServiceControl) Get() (*GravityControllerService, error) {
+	return toServiceConfig(r.svc), nil
+}
+
+// Update updates the controller service.
+func (r *mockServiceControl) Update(config *GravityControllerService) error {
+	if r.svc == nil {
+		r.svc = newService(config)
+		return nil
+	}
+
+	if !shouldUpdate(toServiceConfig(r.svc), config) {
```

Comment:
I don't think additional test functionality (esp. in context of mocks) is a helpful test - this is not used anywhere but the test and replaces the actual implementation. We should consider implementing this as a test to execute in the context of an actual cluster (eventually, we will be running tests that require a cluster and we already have a bunch of them in the repository) using the actual implementation of the service control and cluster config control implementations.